### ISSUE TITLE
BLE Client Example Does Not Work with ServiceUUIDs that have Leading Zeros.

### ISSUE THREAD

**Issue by thomastech** [nkolban/esp32-snippets#912]
I am new to the library and have run into a odd issue regarding my client application. I have a small BLE "find-me" button that has ServiceUUID "0000ffe0-0000-1000-8000-00805f9b34fb" (confirmed with nrf Connect). 

I've followed the library's BLE_client (as well as Andreas Spiess's) example without success. The onResult() callback always fails even though my button is reported in the scan log. 

I've tried these two declarations without success:
```
static BLEUUID serviceUUID("0000ffe0-0000-1000-8000-00805f9b34fb");
static BLEUUID serviceUUID(BLEUUID("0000ffe0-0000-1000-8000-00805f9b34fb"));
```
Eventually I noticed that the logged advertisedDevice's ServiceUUID appears as ffe0-0000-1000-8000-00805f9b34fb, and it is missing the leading zeros.  So I created a BLE test server in nrf Connect with a ServiceUUID that uses leading FFFF's instead of the 0000. **This works!**  

So it appears leading zeros are a problem. For confirmation, I changed the test server's ServiceUUID to use leading zeros and it fails.

I could use some advice in finding a fix regarding the suppressed leading 0000's. I've been reviewing the library and so far I haven't found any joy. Any advice is appreciated.

- Thomas

**Comment by thomastech** [comment]
I found a workaround using this alternate serviceUUID declaration:
`static BLEUUID serviceUUID(BLEUUID("ffe0"));`

So it appears that only 128 bit UUID's are affected and the alternate 16 bit is OK.  Wish I discovered this several days ago, could have avoided pulling out so much hair.

I'll close this issue ticket since I found a solution; But hopefully a developer investigates and provides a fix to the library.

- Thomas

**Comment by thomastech** [comment]
I'll leave this issue open so that a developer can more easily find it. Feel free to close it at any time.
- Thomas

**Comment by thomastech** [comment]
I found the problem and have a fix for it.  In file BLEUUID.cpp, replace the toString() function with this updated version (changes to two lines):

```
/**
 * @brief Get a string representation of the UUID.
 *
 * The format of a string is:
 * 01234567 8901 2345 6789 012345678901
 * 0000180d-0000-1000-8000-00805f9b34fb
 * 0 1 2 3  4 5  6 7  8 9  0 1 2 3 4 5
 *
 * @return A string representation of the UUID.
 */
std::string BLEUUID::toString() {
	if (!m_valueSet) return "<NULL>";   // If we have no value, nothing to format.
	// If the UUIDs are 16 or 32 bit, pad correctly.

	if (m_uuid.len == ESP_UUID_LEN_16) {  // If the UUID is 16bit, pad correctly.
		char hex[9];
		snprintf(hex, sizeof(hex), "0000%04x", m_uuid.uuid.uuid16);
		return std::string(hex) + "-0000-1000-8000-00805f9b34fb";
	} // End 16bit UUID

	if (m_uuid.len == ESP_UUID_LEN_32) {  // If the UUID is 32bit, pad correctly.
		char hex[9];
		snprintf(hex, sizeof(hex), "%08x", m_uuid.uuid.uuid32);
		return std::string(hex) + "-0000-1000-8000-00805f9b34fb";
	} // End 32bit UUID

	// The UUID is not 16bit or 32bit which means that it is 128bit.
	//
	// UUID string format:
	// AABBCCDD-EEFF-GGHH-IIJJ-KKLLMMNNOOPP
	auto size = 37; 
	char *hex = (char *)malloc(size);
	snprintf(hex, size, "%02x%02x%02x%02x-%02x%02x-%02x%02x-%02x%02x-%02x%02x%02x%02x%02x%02x",
			m_uuid.uuid.uuid128[15], m_uuid.uuid.uuid128[14],
			m_uuid.uuid.uuid128[13], m_uuid.uuid.uuid128[12],
			m_uuid.uuid.uuid128[11], m_uuid.uuid.uuid128[10],
			m_uuid.uuid.uuid128[9], m_uuid.uuid.uuid128[8],
			m_uuid.uuid.uuid128[7], m_uuid.uuid.uuid128[6],
			m_uuid.uuid.uuid128[5], m_uuid.uuid.uuid128[4],
			m_uuid.uuid.uuid128[3], m_uuid.uuid.uuid128[2],
			m_uuid.uuid.uuid128[1], m_uuid.uuid.uuid128[0]);
	std::string res(hex);
	free(hex);
	return res;
} // toString
 ```
**EDIT:** This is not the latest code to use. See below for details to the BLE library I ended up using.

- Thomas

**Comment by thomastech** [comment]
After solving the UUID issue and obtaining a successful discovery, I experienced ESP crashes when the server device's desired characteristic was found. Dooh!

After digging through some reported github issues I learned that the V1.0.1 BLE library that is used by Platformio IDE is buggy and out of date. I found updated BLE library files that are stored at https://github.com/nkolban/esp32-snippets/tree/master/cpp_utils that I copied to the built-in library folder. I don't know if these are the latest, but they eliminate the crashes.

So far the revised library has been a good friend. I can now connect to my BLE find-me FOB and detect its button presses.  I may have lost a few days of coding while digging through all this, but the results are fantastic! 

I have to thank @nkolban for sharing his BLE library and the developer(s) that made critical revisions.

- Thomas
